mlrun/datastore/base.py

```diff
@@ -157,6 +158,193 @@ def upload(self, key, src_path):
+            start_time,
+            end_time,
+            time_partitioning_granularity: str,
+            filesystem,
+        ):
+            """
+            Build a list of URLs based on detected partitioning and time range.
+
+            Automatically adjusts iteration step based on deepest partition key.
+            """
+            # Determine smallest partition granularity
+            actual_partition_keys = []
+            for key in ["year", "month", "day", "hour"]:
+                if key in partition_keys:
+                    actual_partition_keys.append(key)
+                    if key == time_partitioning_granularity:
+                        break
+
+            if "hour" in actual_partition_keys:
+                step = datetime.timedelta(hours=1)
+                fmt_keys = ["year", "month", "day", "hour"]
+            elif "day" in actual_partition_keys:
+                step = datetime.timedelta(days=1)
+                fmt_keys = ["year", "month", "day"]
+            elif "month" in actual_partition_keys:
+                # Approximate month stepping
+                fmt_keys = ["year", "month"]
+
+                def next_step(dt):
+                    year = dt.year + (dt.month // 12)
+                    month = 1 if dt.month == 12 else dt.month + 1
+                    return dt.replace(year=year, month=month, day=1)
+            elif "year" in actual_partition_keys:
+                fmt_keys = ["year"]
+
+                def next_step(dt):
+                    return dt.replace(year=dt.year + 1, month=1, day=1)
+            else:
+                raise ValueError("No recognized time-based partition key found")
+
+            urls = []
+            current = start_time
+
+            while current <= end_time:
+                # Build directory structure from partition keys
+                parts = []
+                exists = True
+                for key in fmt_keys:
+                    if key == "year":
+                        parts.append(f"year={current.year}")
+                    elif key == "month":
+                        parts.append(f"month={current.month:02d}")
+                    elif key == "day":
+                        parts.append(f"day={current.day:02d}")
+                    elif key == "hour":
+                        parts.append(f"hour={current.hour:02d}")
+
+                    if parts and not filesystem.isdir(
+                        f"{base_path.rstrip('/')}/" + "/".join(parts)
+                    ):
+                        exists = False
+                        break
+
+                if exists:
+                    urls.append(f"{base_path.rstrip('/')}/" + "/".join(parts))
+
+                # Advance to next period
+                if exists or key == time_partitioning_granularity:
+                    if (
+                        "hour" in actual_partition_keys
+                        or "day" in actual_partition_keys
+                    ):
```

Comment:
```suggestion
                    if step:
````
and set `step=None` at the start. Makes it more obviously correct to humans as well as to static analysis tools like PyCharm. And saves a few cycles in the process.